test/cloudpassage_lib/scans_test.clj

```
@@ -163,28 +167,113 @@
            ms/stream->seq
            doall))))
 
+(defn ^:private make-fake-page
+  "Generates a fake page function. Give it the expected base path and a
+   function to call with the current page number."
+  [cb]
+  (fn [client-id client-secret url]
+    (is (= client-id "lvh"))
+    (is (= client-secret "hunter2"))
+    (let [parsed-url (u/url url)
+          path (:path parsed-url)
+          query (:query parsed-url)
+          page-num (-> query
+                       (get "page" "0")
+                       Integer/parseInt)]
+      (cb {:page-num page-num :path path}))))
+
+(defn fake-servers-page
+  "Pretends to be the servers list page results."
+  [page-two-breaks]
+  (make-fake-page (fn [{:keys [page-num path]}]
+                    (is (= "/v1/servers" path))
+                    (case page-num
+                      0 {:servers [{:id "server-id-1"} {:id "server-id-2"}]
+                         :pagination {:next (str @#'scans/base-servers-url "?page=1")}}
+                      1 (if page-two-breaks
+                          :cloudpassage-lib.core/fetch-error
+                          {:servers [{:id "server-id-3"}]})))))
+
+(deftest list-servers!-tests
+  (testing "Returns all servers if paginated call is OK."
+    (with-redefs [scans/get-page! (fake-servers-page false)]
+      (let [server-stream (scans/list-servers! "lvh" "hunter2")
+            id-list (map :id (ms/stream->seq server-stream))]
+        (is (= ["server-id-1" "server-id-2" "server-id-3"] id-list))
+        (is (ms/closed? server-stream)))))
+  (testing "If page 2 is bad, a partial list is returned."
+    (with-redefs [scans/get-page! (fake-servers-page true)]
+      (let [server-stream (scans/list-servers! "lvh" "hunter2")
+            id-list (map :id (ms/stream->seq server-stream))]
+        (is (= ["server-id-1" "server-id-2"] id-list))
```

Is this the behaviour we want? If we can't fetch all the pages, we won't be able to return a list of all the servers, so we probably want to throw an exception and die. I'm pretty sure that's what the code says we do, but it looks like we're not actually exercising that logic here.